fix(ivy): incorrect namespace for root node created through ViewContainerRef

crisbeto opened the pull request:
Currently in Ivy whenever we encounter a new namespace, we set it in the global state so that all subsequent nodes are created under the same namespace. Next time a template is run the namespace will be reset back to HTML.

This breaks down if the last node that was rendered was under the SVG or MathML namespace and we create a component through `ViewContainerRef.create`, because the next template function hasn't run yet and it hasn't had the chance to update the namespace. The result is that the root node of the new component will retain the wrong namespace and may not end up rendering at all (e.g. if we're trying to show a `div` inside the SVG namespace). This issue has the potential to affect a lot of apps, because all components inserted through the router also go through `ViewContainerRef.create`.

pkozlowski-opensource commented on packages/core/src/render3/component_ref.ts:
```diff
@@ -140,6 +140,9 @@ export class ComponentFactory<T> extends viewEngine_ComponentFactory<T> {
 
+    // Ensure that the namespace for the root node is correct,
+    // otherwise the browser might not render out the element properly.
+    ɵɵnamespaceHTML();
```
Could we avoid calling instructions here?

crisbeto replied:
Do we have an alternative for it? I followed what we were doing inside `executeTemplate`.

kara replied:
I think what he means is that we should create internal versions of the instructions rather than referencing them directly in non-generated code.

```suggestion
    namespaceHTMLInternal();
```

kara reviewed (approved):
LGTM, once Pawel's comment is addressed

crisbeto commented:
Changed based on the feedback.

kara commented on packages/core/src/render3/state.ts:
```diff
@@ -540,6 +540,14 @@ export function ɵɵnamespaceMathML() {
+}
+
+/**
+ * Sets the namespace used to create elements no `null`, which forces element creation to use
```
```suggestion
 * Sets the namespace used to create elements to `null`, which forces element creation to use
```

crisbeto commented:
The Codefresh failure looks like a flake.

crisbeto committed:
fix(ivy): incorrect namespace for root node created through ViewContainerRef

Currently in Ivy whenever we encounter a new namespace, we set it in the global state so that all subsequent nodes are created under the same namespace. Next time a template is run the namespace will be reset back to HTML.

This breaks down if the last node that was rendered was under the SVG or MathML namespace and we create a component through `ViewContainerRef.create`, because the next template function hasn't run yet and it hasn't had the chance to update the namespace. The result is that the root node of the new component will retain the wrong namespace and may not end up rendering at all (e.g. if we're trying to show a `div` inside the SVG namespace). This issue has the potential to affect a lot of apps, because all components inserted through the router also go through `ViewContainerRef.create`.

kara commented:
[VE presubmit](http://test/OCL:255256719:BASE:255256737:1561582226967:467aa371)

[Ivy presubmit](http://test/OCL:255256719:BASE:255257039:1561582322095:2f860cab)
